core/kernel/tcg.c

```diff
@@ -0,0 +1,399 @@
+static uint32_t tcg_event_final_size(struct tpml_digest_values *digest_list)
+{
+	uint32_t len = 0;
+	size_t i = 0;
+
+	len = offsetof(struct tcg_pcr_event2, digests);
+	len += offsetof(struct tpml_digest_values, digests);
+	for (i = 0; i < digest_list->count; i++) {
+		uint16_t hash_alg = digest_list->digests[i].hash_alg;
+
+		len += offsetof(struct tpmt_ha, digest);
+		len += tpm2_get_alg_len(hash_alg);
+	}
+	len += sizeof(uint32_t); /* tcg_pcr_event2 event_size*/
+
+	return len;
+}
+
+/*
+ * tcg_parse_event_log_header() -  Parse and verify the event log header fields
+ *
+ * @buffer:			Pointer to the start of the eventlog
+ * @size:			Size of the eventlog
+ * @pos:			Return offset of the next event in buffer right
+ *				after the event header i.e specID
+ *
+ * Return:	status code
+ */
+static TEE_Result tcg_parse_event_log_header(void *buffer, uint32_t size,
+					     uint32_t *pos)
+{
+	struct tcg_pcr_event *event_header = (struct tcg_pcr_event *)buffer;
+	uint32_t i = 0;
+
+	if (size < sizeof(*event_header))
+		return TEE_ERROR_BAD_FORMAT;
+
+	if (get_unaligned_le32(&event_header->pcr_index) != 0 ||
+	    get_unaligned_le32(&event_header->event_type) != EV_NO_ACTION)
+		return TEE_ERROR_BAD_FORMAT;
+
+	for (i = 0; i < sizeof(event_header->digest); i++) {
+		if (event_header->digest[i])
+			return TEE_ERROR_BAD_FORMAT;
+	}
+
+	*pos += sizeof(*event_header);
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * tcg_parse_specid_event() -  Parse and verify the specID Event in the eventlog
+ *
+ * @buffer:		Pointer to the start of the eventlog
+ * @log_size:		Size of the eventlog
+ * @pos:		[in] Offset of specID event in the eventlog buffer
+ *			[out] Return offset of the next event in the buffer
+ *			after the specID
+ * @digest_list:	list of digests in the event
+ *
+ * Return:		status code
+ */
+static TEE_Result tcg_parse_specid_event(void *buffer, uint32_t log_size,
+					 uint32_t *pos,
+					 struct tpml_digest_values *digest_list)
+{
+	struct tcg_efi_spec_id_event *spec_event = NULL;
+	struct tcg_pcr_event *event_header = buffer;
+	uint8_t vendor_sz = 0;
+	uint16_t hash_alg = 0;
+	uint32_t active = 0;
+	uint32_t alg_count = 0;
+	uint32_t i = 0;
+	uint32_t pcr_count = 0;
+	uint32_t spec_active = 0;
+	uint32_t supported = 0;
+	size_t spec_event_size = 0;
+
+	if (*pos >= log_size || (*pos + sizeof(*spec_event)) > log_size)
+		return TEE_ERROR_BAD_FORMAT;
+
+	/* Check specID event data */
+	spec_event = (struct tcg_efi_spec_id_event *)((uintptr_t)buffer + *pos);
+	/* Check for signature */
+	if (memcmp(spec_event->signature, TCG_EFI_SPEC_ID_EVENT_SIGNATURE_03,
+		   sizeof(TCG_EFI_SPEC_ID_EVENT_SIGNATURE_03))) {
+		EMSG("specID Event: Signature mismatch");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	if (spec_event->spec_version_minor !=
+			TCG_EFI_SPEC_ID_EVENT_SPEC_VERSION_MINOR_TPM2 ||
+	    spec_event->spec_version_major !=
+			TCG_EFI_SPEC_ID_EVENT_SPEC_VERSION_MAJOR_TPM2)
+		return TEE_ERROR_BAD_FORMAT;
+
+	if (spec_event->number_of_algorithms < 1) {
```

Comment:
Looks like a complicated way of testing for zero.

Reply:
Changed it